packages/eslint-plugin/docs/rules/typedef.md

```diff
@@ -0,0 +1,364 @@
+# Require type annotations to exist (typedef)
+
+TypeScript cannot always infer types for all places in code.
+Some locations require type annotations for their types to be inferred.
+
+```ts
+class ContainsText {
+  // There must be a type annotation here to infer the type
+  delayedText: string;
+
+  // `typedef` requires a type annotation here to maintain consistency
+  immediateTextExplicit: string = 'text';
+
+  // This is still a string type because of its initial value
+  immediateTextImplicit = 'text';
+}
+```
+
+> Note: requiring type annotations unnecessarily can be cumbersome to maintain and generally reduces code readability.
+> TypeScript is often better at inferring types than easily written type annotations would allow.
+> Instead of enabling `typedef`, it is generally recommended to use the `--noImplicitAny` and/or `--strictPropertyInitialization` compiler options to enforce type annotations only when useful.
+
+## Rule Details
+
+This rule can enforce type annotations in locations regardless of whether they're required.
+This is typically used to maintain consistency for element types that sometimes require them.
+
+## Options
+
+This rule has an object option that may receive any of the following as booleans:
+
+- `"arrayDestructuring"`
+- `"arrowCallSignature"`: `true` by default
+- `"arrowParameter"`: `true` by default
+- `"callSignature"`: `true` by default
+- `"memberVariableDeclaration"`: `true` by default
+- `"objectDestructuring"`
+- `"parameter"`: `true` by default
+- `"propertyDeclaration"`: `true` by default
+- `"variableDeclaration"`
+
+For example, with the following configuration:
+
+```json
+{
+  "rules": {
+    "typedef": [
+      "error",
+      {
+        "arrowParameter": false,
+        "variableDeclaration": true
+      }
+    ]
+  }
+}
+```
+
+- Type annotations on arrow function parameters are not required
+- Type annotations on variables are required
+- Options otherwise adhere to the defaults
+
+### arrayDestructuring
+
+Whether to enforce type annotations on variables declared using array destructuring.
+
+Examples of **incorrect** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a] = [1];
+const [b, c] = [1, 2];
+```
+
+Examples of **correct** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a]: number[] = [1];
+const [b]: [number] = [2];
+const [c, d]: [boolean, string] = [true, 'text'];
+```
+
+### arrowCallSignature
+
+Whether to enforce type annotations for return types of arrow functions.
```

Comment:
is this option worthwhile when we have `explicit-function-return-type`?
It seems like duplicated work?

Reply:
🤷‍♂ it existed in the old rule, so I ported it here. That brings up the question: why have `explicit-function-return-type` when `typedef` can do the same thing? Should `typedef` be split up into separate rules?

Reply:
`explicit-function-return-type` goes one step further, with options for the user to control how much type inference they want to allow for their return types.

examples
- you can allow not including a return type if the function is in a place which has a function type with a return type `const x: () => number = () => 1`
- there's a PR right now to allow higher order functions (`() => () => () => (): number => 1`)

Reply:
Nifty. I'm a little confused - @bradzacher what would you prefer this PR do? I'm equally happy...

* ...changing this option in some way?
* ...removing this option?
* ...splitting `typedef` into granular rules with names like `explicit-*-type`?

Reply:
I would probably look to removing the two options that conflict with `explicit-function-return-type` (`arrowCallSignature` and `callSignature`).

Just to help people migrate, might be good to add a note to the docs directing people to `explicit-function-return-type`.

I'm happy with the rest of the options being in here.
I don't see a need to spend the time / add the complexity of adding one rule per option unless someone decides there needs to be options for specific cases.